src/api/c++/z3++.h

```diff
@@ -392,21 +393,20 @@ namespace z3 {
 
     template<typename T>
     class array {
-        T *      m_array;
+        std::unique_ptr<T[]> m_array;
         unsigned m_size;
-        array(array const & s);
-        array & operator=(array const & s);
+        array(array const & s) = delete;
+        array & operator=(array const & s) = delete;
     public:
-        array(unsigned sz):m_size(sz) { m_array = new T[sz]; }
+        array(unsigned sz):m_array(new T[sz]),m_size(sz) {}
         template<typename T2>
         array(ast_vector_tpl<T2> const & v);
-        ~array() { delete[] m_array; }
-        void resize(unsigned sz) { delete[] m_array; m_size = sz; m_array = new T[sz]; }
+        void resize(unsigned sz) { m_array.reset(new T[sz]); m_size = sz; }
         unsigned size() const { return m_size; }
         T & operator[](int i) { assert(0 <= i); assert(static_cast<unsigned>(i) < m_size); return m_array[i]; }
         T const & operator[](int i) const { assert(0 <= i); assert(static_cast<unsigned>(i) < m_size); return m_array[i]; }
-        T const * ptr() const { return m_array; }
-        T * ptr() { return m_array; }
+        T const * ptr() const { return & m_array[0]; }
+        T * ptr() { return & m_array[0]; }
```

Comment:
I have seen debug runtimes assert on these kinds of tricks. 
Introduces maintenance hassles. Is this really best practice to use &v[0]?

Reply:
Don't you want https://en.cppreference.com/w/cpp/memory/unique_ptr/get

Reply:
According to the docs, the array version does not support raw pointer stuff, so this is how it has to be done. Here:
```
Single-object version, unique_ptr<T>
operator*operator-> | dereferences pointer to the managed object(public member function)
Array version, unique_ptr<T[]>
operator[] | provides indexed access to the managed array
```
Source: https://en.cppreference.com/w/cpp/memory/unique_ptr

Reply:
Regardless, since it is a smart pointer, the best we could reduce this to would be: `&*m_array`, if it did support `*` in array mode.

Reply:
I can update it to `get`.